web/server/go.md

```diff
@@ -0,0 +1,498 @@
+        cutenessScore := nameAsInt%10 + BRONT_ADDER
+        return cutenessScore, nil
+    }
+}
+```
+
+So, we've added a new implementation for cuteness score
+to the `services` package.
+It returns a function that can calculate cuteness for a given dog ID.
+The dependencies, like the `Database` connector,
+are passed into the function generator and hidden from the return value.
+
+Notice how this definition is independent of the definition in the handler
+(in a different package).
+The handler only needs to know what is required to return a score,
+it does *not* need to know how.
+Developing the implementation independently
+allows us to easily swap out function implementations
+if the user, client, or dependency requirements ever change.
+For example,
+what if the cuteness score changes to be out of 100 instead of 10?
+Instead of dealing with that in a handler,
+which we've defined as being responsible for request/response satisfaction,
+it's segmented to the service function/package.
+Likewise, if we want to use a new DB connector,
+the dependency for the service will change,
+but the handler will be agnostic to that difference.
+If we reuse the calculator elsewhere,
+changes will not have to be updated sparsely across the handlers codebase.
+
+#### Testing Services
+
+The focus on testing services
+should be unit testing business logic.
+These tests will likely follow feature workflows.
+An example from above would be:
+given a dog with X characters in its name,
+what is its dog score?
+This layer is ripe for testing with
+[table driven tests](https://github.com/golang/go/wiki/TableDrivenTests)
+(ex. given this array of dog names,
+what are their scores).
+
+Avoid over-testing dependencies,
+and attempt to mock them when possible
+(such as DB calls),
+especially expensive or unknown resources such as third party APIs.
+These are likely to create long
+and flaky tests.
+
+#### Service Resources
+
+There  are three patterns worth understanding are:
+
+1. [Service Objects in Go](https://www.calhoun.io/using-the-service-object-pattern-in-go/).
+   This demonstrates the origin of the pattern,
+   where single function structs/interfaces
+   are used to encapsulate work.
+2. [How to Use Go Interface](https://blog.chewxy.com/2018/03/18/golang-interfaces/).
+   This guide will provide a "Go" styling to interfaces.
+   Rather than use global objects (like in Ruby)
+   or tightly coupled interfaces/classes (Java),
+   we'll keep interfaces close to usage (the handler in this case),
+   and implementations will implicitly satisfy those contracts.
+3. [What is a Closure](https://www.calhoun.io/what-is-a-closure/).
+   Taking the "one function interface" pattern a step further,
+   we can utilize Go's first class functions and closures
+   to simply pass the function rather than struct,
+   removing a layer of indirection.
+
+### General Patterns
+
+This section is for patterns that don't fit into the specific layers defined above
```

Comment:
Filling this out is a TODO. Was going to fill this in, but the doc ended up being larger than I thought. Happy to delete for now.

Reply:
I'd suggest something along the lines of this (via https://peter.bourgon.org/go-best-practices-2016/): 
> Loggers are dependencies, just like references to other components, database handles, commandline flags, etc.

I've been a huge fan of the go-kit log package and the patterns it espouses... But I think those patterns may be useful/applied to other logging packages that are more performant. (Tho, that may be premature optimization.)